### PR TITLE
ChannelOutboundBuffer bytes before writable accessor

### PR DESCRIPTION
**Pull request by Scottmitch**
Motiviation:
There are currently no accessors which provide visbility into how many bytes must be written in order for a writability change to occur. This feature would be useful for codecs which intent to control how many bytes are queued at any given time.

Modifications:
- add bytesBeforeUnWritable() which will give the number of bytes before the buffer (and associated channel) transitions to not writable
- add bytesBeforeWritable() which will give the number of bytes that must be drained from the queue until the channel becomes writable.

Result:
More visibility into writability for the ChannelOutboundBuffer.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler @ejona86 - FYI. Synchronization may need review...and this method may be a "best" effort basis if multiple threads are in play.  For HTTP/2 I'm not sure if this is an issue because I believe we are assuming writes are done on the event loop thread.

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 1]
```diff
@@ -690,6 +690,36 @@ public long totalPendingWriteBytes() {
...
+            return isWritable() ? bytes : 0;
+        }
+        return 0;
+    }
+
+    /**
+     * Get how many bytes must be drained from the underlying buffer until {@link #isWritable()} returns {@code true}.
+     * This quantity will always be non-negative. If {@link #isWritable()} is {@code true} then 0.
+     */
+    public long bytesBeforeWritable() {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: mark both methods final ?

**Comment by normanmaurer** [comment]
@Scottmitch just a small comment

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 1 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: done.

**Comment by nmittler** [comment]
@Scottmitch Looks great! I think the synchronization is fine ... this is just to give us a ballpark anyway.  And since most of the operations on the buffer will be from the same thread, it's even less risky.

**Comment by Scottmitch** [comment]
@nmittler - +1. I'll also pull together a PR to put similar methods on the `Channel`.

@normanmaurer - Turns out `ChannelOutboundBuffer` is final...so I can't make the methods final...reverting change.

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 1 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: Turns out ChannelOutboundBuffer is final...so I can't make the methods final...reverting change.

**Comment by normanmaurer** [comment]
@Scottmitch argh... right!

**Comment by Scottmitch** [comment]
Cherry-picked 4.0 (14da966) 4.1 (43a558b) master (ffdf00e)

**Pull request closed without merging**